pkg/controllers/oauthendpoints/oauth_endpoints_controller.go

```diff
@@ -0,0 +1,273 @@
+	}
+	for _, port := range service.Spec.Ports {
+		results = append(results, net.JoinHostPort(service.Spec.ClusterIP, strconv.Itoa(int(port.Port))))
+	}
+	if len(results) == 0 {
+		return nil, fmt.Errorf("no valid oauth services found")
+	}
+	return toHealthzURL(results), nil
+}
+
+func listOAuthRoutes(routeLister routev1listers.RouteLister, recorder events.Recorder) ([]string, error) {
+	var results []string
+	route, err := routeLister.Routes("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		recorder.Warningf("OAuthRouteCheck", "failed to get oauth route: %v", err)
+		return nil, err
+	}
+	for _, ingress := range route.Status.Ingress {
+		if len(ingress.Host) > 0 {
+			for _, condition := range ingress.Conditions {
+				if condition.Type == routev1.RouteAdmitted && condition.Status == corev1.ConditionTrue {
+					results = append(results, ingress.Host)
+					break
+				}
+			}
+
+		}
+	}
+	if len(results) == 0 {
+		recorder.Warningf("OAuthRouteCheck", "route status does not have host address")
+		return nil, fmt.Errorf("route status does not have host address")
+	}
+	return toHealthzURL(results), nil
+}
+
+func getOAuthRouteTLSConfig(cmLister corev1listers.ConfigMapLister, secretLister corev1listers.SecretLister, ingressLister configv1lister.IngressLister, systemCABundle []byte, recorder events.Recorder) (*tls.Config, error) {
+	// get default router CA cert cm
+	defaultIngressCertCM, err := cmLister.ConfigMaps("openshift-config-managed").Get("default-ingress-cert")
+	if err != nil {
+		recorder.Warningf("OAuthRouterCACerts", "failed to retrieve the default router CA certs: %v", err)
+		return nil, err
+	}
+
+	defaultRouterCAPEM, ok := defaultIngressCertCM.Data["ca-bundle.crt"]
+	if !ok {
+		klog.Info("the openshift-config-managed/default-ingress-cert CM does not contain the \"ca-bundle.crt\" key")
+		return nil, err
+	}
+
+	ingress, err := ingressLister.Get("cluster")
+	if err != nil {
+		recorder.Warningf("OAuthRouteSecret", "failed to get ingress config: %v", err)
+		return nil, err
+	}
+	if len(ingress.Spec.Domain) == 0 {
+		return nil, fmt.Errorf("ingress config domain cannot be empty")
+	}
+
+	routerSecret, err := secretLister.Secrets("openshift-authentication").Get("v4-0-config-system-router-certs")
+	if err != nil {
+		recorder.Warningf("OAuthRouteSecret", "failed to get oauth route ca cert: %v", err)
+		return nil, err
+	}
+
+	// find the domain that matches our route
+	routerCertKey, ok := routerSecret.Data[ingress.Spec.Domain]
+	if !ok {
+		klog.Infof("unable to find router certs for domain %s", ingress.Spec.Domain)
+		return nil, nil
+	}
+
+	rootCAs := x509.NewCertPool()
+	if ok := rootCAs.AppendCertsFromPEM([]byte(defaultRouterCAPEM)); !ok {
+		klog.Infof("the default ingress CA bundle did not contain any PEM certificates %s", defaultRouterCAPEM)
+		return nil, nil
+	}
+
+	if ok := rootCAs.AppendCertsFromPEM(routerCertKey); !ok {
+		klog.Infof("failed to parse router certs for domain %s", ingress.Spec.Domain)
```

Comment:
same here

Reply:
It's usually not a parsing error, just none certificates are present in the bundle. I'll improve the message.